### PR TITLE
[Model] Optimize MiniMax-H3 DLO component lifecycle

### PR DESCRIPTION
**Pull request by hsliuustc0106**
## Summary

- Retain reusable allocator blocks across MiniMax-H3 DLO encoder → DiT → VAE component boundaries while cached memory stays within 25% of device capacity and at least 5% remains physically free.
- Generalize `PinnedModuleStager` to stage one immutable module group with a shared copy stream/event while preserving parameter and buffer storage aliases, shapes, strides, offsets, and dtypes.
- Stage the encoder's non-block children from that immutable pinned CPU master, eliminating their GPU→CPU weight rematerialization.
- Force cache release and restore host masters on component/staging failure, missing telemetry, or an OOM retry. The ordinary DiT prefetch path is unchanged.

This remains a focused lifecycle change. Host Weight Runtime is now present on `main`, but this PR does not alter its selection, lease, staging-slot, or restore contracts.

## Review follow-up

Head `1843f05c` is rebased onto current `main` `29e3165b` and addresses all recent inline comments:

- `BaseException` during failure-path offload/cache cleanup is contained so forced release still runs and the original component failure is preserved.
- A repeated-lifecycle regression test covers `offload_to_cpu()` failing on the success path, forced cache release, and re-raise.
- The feature and module design docs now specify the 25% cached / 5% physical-free bounds, conservative telemetry fallback, OOM retry, memory tradeoff, and unchanged executor shutdown cleanup.

## Current matched L20X A/B

Both sides ran under one `gpu run` reservation on the same NVIDIA L20X (143,771 MiB, physical GPU 5) with fixed `cpunodebind=1,membind=0`, model/cache paths, prompt, seeds, and environment: Python 3.12.13, PyTorch 2.13.0+cu129, CUDA 12.9, driver 570.133.20, vLLM 0.27.0, and vLLM-Omni source at the listed SHAs.

Configuration: DP1/TP1, 1344×768, 5 seconds, 24 fps, `num_inference_steps=2` (one distilled denoising pass), `CUDNN_ATTN`, no DLO AllGather, zero resident layers, one excluded warmup, and three measured requests.

| Unprofiled metric | `main` (`29e3165b`) | This PR (`1843f05c`) | Delta |
|---|---:|---:|---:|
| Request latency, mean ± population stddev | 18.317 ± 0.267 s | 15.580 ± 0.032 s | **-14.94%** |
| Three measured requests | 18.458 / 18.549 / 17.943 s | 15.619 / 15.541 / 15.579 s | — |
| Peak HBM | 20,808 MB | 22,722 MB | +1,914 MB (+9.20%) |
| Reserved memory after requests | 7.443–7.464 GB | 23.826 GB on all three | retained-cache plateau |

Preparation is reported separately: process-to-ready was 114.930 s for base and 89.665 s for head because the base run populated shared filesystem/page caches. Startup is not used in the latency comparison.

The same-shape no-offload reference peaks at 133,298 MB, so optimized DLO still reduces peak HBM by **82.95%** (110,576 MB). Throughput/concurrency is not claimed.

### Output parity

Base and head used the same prompt and seed for the final measured request and were byte-identical:

- video `[124, 768, 1344, 3]`, float32: `2c4b0ce4b20743a54eece09d08d3b3891aeb480c49dd14ecf6495d63a27571e6`
- audio `[1, 2, 165600]`, float32: `4cd6f8e56e82333f3fdf758288b58f9784d4725dbe55f8e93ba53a33b22a9170`

Both configurations shut down with zero active child processes, and GPU 5 returned to 0 MB used.

## Nsight mechanism evidence

The matched pre-rebase development capture (same lifecycle implementation, base `1c917b27`, head `dd130e17`) remains the detailed allocator/transfer trace. Profiled timings are separate from the unprofiled A/B above.

| Captured metric | Base | Optimized |
|---|---:|---:|
| Instrumented stage / NVTX GPU span | 18.823 / 18.812 s | 16.827 / 16.773 s |
| Kernel count / union | 311,426 / 12.160 s | 311,426 / 12.533 s |
| H2D payload / device time | 130.856 GB / 2.907 s | 130.856 GB / 2.398 s |
| H2D overlapped / exposed | 1.338 / 1.569 s | 1.199 / 1.198 s |
| D2H count / payload | 69 / 3.461 GB | 41 / 1.537 GB |
| GPU busy / idle in active window | 14.775 / 4.037 s | 15.148 / 1.625 s |
| `cudaMalloc` / `cudaFree` | 150 / 150 | **1 / 0** |

The 1.923 GB D2H payload reduction is the removed encoder non-block round trip. The policy added three `cudaMemGetInfo` calls totaling 1.6 ms. The outer profile wall time includes `cudaProfilerStop` report materialization and is not used.

## Validation

- Focused review suite: **11 passed**, 14 warnings.
- Broader CPU offloader/MiniMax suite: **339 passed, 14 deselected**, 15 warnings.
- Local pre-commit: all changed-file hooks passed, including Ruff, mypy, SPDX, test markers, markdownlint, forbidden imports, and the direct-`torch.cuda` guard.
- Exact DP2/TP1 no-AllGather smoke at 1344×768 returned video `[124,768,1344,3]` and audio `[1,2,165600]`, peaked at 21,256 MB, and shut down with zero active children.
- Independent B300 validation on the original head reported 291 CPU tests passing plus a successful exact-resolution lifecycle smoke and clean shutdown; it was functional validation, not a B300 performance A/B.
- InferMatrixCopilot re-reviewed final head `1843f05c`; all three prior findings were rechecked as fixed and no new finding remained.

Checked-in functional command:

```bash
gpu run --gpus 1 -- env VLLM_WORKER_MULTIPROC_METHOD=spawn VLLM_OMNI_VIDEO_SYNC_TIMEOUT=14400 python examples/offline_inference/minimax_h3/dlo_lifecycle.py --model /path/to/MiniMax-H3/FL2VA --mode dlo-no-allgather --dp-size 1 --steps 2 --duration 5 --width 1344 --height 768
```

Review-fix A/B artifacts and the non-production measurement harness are under `/home/hsliu/tmp/pr6526-reviewfix-20260825/`; earlier profiles remain under `/home/hsliu/tmp/dlo-lifecycle-final-validation/`.

## Scope and limitations

- The extra 1,914 MB peak HBM is the intentional latency tradeoff. On smaller or busy devices, crossing either bound returns to ordinary per-component cache release, so the speedup may be smaller.
- No encoder residency/quantization, VAE prefetch, DiT transfer/prefetch, or HWR behavior change is included.
- AllGather performance and non-CUDA performance were not re-benchmarked; other MiniMax-H3 DLO topologies/platforms receive the bounded policy but are not performance claims here.
- This optimizes single-request lifecycle latency and does not change current no-AllGather admission or claim DP throughput scaling.

### CODE REVIEW

**Comment by lishunyang12** [comment]
## Independent test result

Tested exact head `dd130e1700c92d8b01209de861accd4683bac1e8` on 2026-08-24.

Environment: NVIDIA B300 SXM6 AC, Python 3.12.3, PyTorch 2.13.0+cu130, CUDA 13.0, vLLM 0.27.0, vLLM-Omni 0.27.0.dev106.

### Unit and regression tests

```bash
python -m pytest -o addopts='' -q \
  tests/diffusion/models/minimax_h3/test_minimax_h3_dlo_lifecycle.py \
  tests/diffusion/offloader/test_module_residency.py
```

- **10 passed**, 17 warnings.

```bash
python -m pytest -o addopts='' -q -m cpu \
  tests/diffusion/offloader tests/diffusion/models/minimax_h3 \
  --ignore=tests/diffusion/models/minimax_h3/test_minimax_h3_e2e.py
```

- **291 passed, 14 deselected**, 18 warnings.

### B300 DLO lifecycle smoke

```bash
CUDA_VISIBLE_DEVICES=7 \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
VLLM_OMNI_VIDEO_SYNC_TIMEOUT=14400 \
python examples/offline_inference/minimax_h3/dlo_lifecycle.py \
  --model /path/to/MiniMax-H3/FL2VA \
  --mode dlo-no-allgather --dp-size 1 \
  --steps 2 --duration 5 --width 1344 --height 768
```

Result:

- video shape: `[124, 768, 1344, 3]`
- audio shape: `[1, 2, 165600]`
- worker peak memory: **21,256 MB**
- valid request: **41.291 s** (`stage_0_gen_ms=41,289.444`)
- shutdown: **2.581 s**
- active child processes after shutdown: `[]`

The functional lifecycle smoke passed and released all GPU/worker processes cleanly. This independently validates the affected behavior on B300; it does not attempt to reproduce the PR's L20X latency A/B claim. All current GitHub checks are also green.

**Comment by hsliuustc0106** [comment]
> ## Independent test result
> Tested exact head `dd130e1700c92d8b01209de861accd4683bac1e8` on 2026-08-24.
> 
> Environment: NVIDIA B300 SXM6 AC, Python 3.12.3, PyTorch 2.13.0+cu130, CUDA 13.0, vLLM 0.27.0, vLLM-Omni 0.27.0.dev106.
> 
> ### Unit and regression tests
> ```shell
> python -m pytest -o addopts='' -q \
>   tests/diffusion/models/minimax_h3/test_minimax_h3_dlo_lifecycle.py \
>   tests/diffusion/offloader/test_module_residency.py
> ```
> 
> * **10 passed**, 17 warnings.
> 
> ```shell
> python -m pytest -o addopts='' -q -m cpu \
>   tests/diffusion/offloader tests/diffusion/models/minimax_h3 \
>   --ignore=tests/diffusion/models/minimax_h3/test_minimax_h3_e2e.py
> ```
> 
> * **291 passed, 14 deselected**, 18 warnings.
> 
> ### B300 DLO lifecycle smoke
> ```shell
> CUDA_VISIBLE_DEVICES=7 \
> VLLM_WORKER_MULTIPROC_METHOD=spawn \
> VLLM_OMNI_VIDEO_SYNC_TIMEOUT=14400 \
> python examples/offline_inference/minimax_h3/dlo_lifecycle.py \
>   --model /path/to/MiniMax-H3/FL2VA \
>   --mode dlo-no-allgather --dp-size 1 \
>   --steps 2 --duration 5 --width 1344 --height 768
> ```
> 
> Result:
> 
> * video shape: `[124, 768, 1344, 3]`
> * audio shape: `[1, 2, 165600]`
> * worker peak memory: **21,256 MB**
> * valid request: **41.291 s** (`stage_0_gen_ms=41,289.444`)
> * shutdown: **2.581 s**
> * active child processes after shutdown: `[]`
> 
> The functional lifecycle smoke passed and released all GPU/worker processes cleanly. This independently validates the affected behavior on B300; it does not attempt to reproduce the PR's L20X latency A/B claim. All current GitHub checks are also green.

did you see the optimization with A/B test?

**Comment by lishunyang12** [comment]
Not yet. My B300 run was a head-only functional lifecycle smoke, not a matched performance A/B, so it validates output correctness, peak memory, and clean shutdown but does not independently verify the reported `-14.08%` L20X latency improvement.

A valid independent comparison would need base `1c917b27` versus head `dd130e17` on the same B300, with identical settings, a warmup, and repeated measured requests. I therefore did not compare my cold B300 `41.291 s` request with the L20X numbers in the PR body or claim that I observed the optimization.

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py [hunk 1]
```diff
@@ -1206,13 +1211,32 @@ def _component_on_device(self, component: nn.Module):
...
         try:
+            if staged:
+                component.load_to_device()
             yield
-        finally:
+        except BaseException:
+            if staged:
+                try:
+                    component.offload_to_cpu()
+                except Exception:
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: P3 (nit): This failure-branch cleanup catches `Exception`, but the success-path branch below catches `BaseException` — a `BaseException` from `offload_to_cpu()` (e.g. KeyboardInterrupt) would skip the forced cache release entirely and displace the original component exception; aligning on `except BaseException` keeps the two branches symmetric.

File: vllm_omni/diffusion/offloader/module_residency.py [hunk 2]
```diff
@@ -1,89 +1,271 @@
...
+    Missing memory telemetry is handled conservatively by releasing the cache.
+    Failure paths can force release; normal executor shutdown keeps its own
+    unconditional device-cache cleanup because this policy is not global.
+    """
+
+    def __init__(
+        self,
+        device: torch.device,
+        *,
+        max_cached_fraction: float = 0.25,
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/offloader/module_residency.py:43]: P3 (docs): The 25% cached / 5% free retention bounds are a user-visible device-memory behavior change (measured +1.9 GB peak; up to 25% of capacity retained between components), but neither `docs/design/feature/offloader/distributed_layerwise_offload.md` nor `docs/design/module/diffusion/offloader.md` mentions the policy — worth a short paragraph on the bounds and the failure-path force release.

File: tests/diffusion/models/minimax_h3/test_minimax_h3_dlo_lifecycle.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    pipeline._dlo_component_cache = mocker.Mock()
+    component = mocker.Mock()
+
+    with pytest.raises(RuntimeError, match="component failed"):
+        with pipeline._component_on_device(component):
+            raise RuntimeError("component failed")
+
+    component.load_to_device.assert_called_once_with()
+    component.offload_to_cpu.assert_called_once_with()
+    pipeline._dlo_component_cache.release_if_needed.assert_called_once_with(force=True)
```
Comment (hsliuustc0106) [inline tests/diffusion/models/minimax_h3/test_minimax_h3_dlo_lifecycle.py:77]: P3 (test): This pins the component-failure branch, but the sibling new behavior — `offload_to_cpu()` itself failing on the success path (`except BaseException` → forced release → re-raise) — has no coverage; a `side_effect` on `offload_to_cpu` that raises the second time would cover it.

File: vllm_omni/diffusion/offloader/module_residency.py (hunk 2 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/offloader/module_residency.py:43]: Documented in 1843f05c in both the DLO feature and offloader module design docs, including the bounds, conservative/forced release, OOM retry, and memory tradeoff.

File: vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py (hunk 1 above)
Comment (hsliuustc0106) [reply vllm_omni/diffusion/models/minimax_h3/pipeline_minimax_h3.py]: Fixed in 1843f05c: failure-path offload and forced-cache cleanup now catch BaseException, preserve the original component error, and have regression coverage.

File: tests/diffusion/models/minimax_h3/test_minimax_h3_dlo_lifecycle.py (hunk 3 above)
Comment (hsliuustc0106) [reply tests/diffusion/models/minimax_h3/test_minimax_h3_dlo_lifecycle.py:77]: Added repeated-lifecycle success-path offload-failure coverage in 1843f05c; it verifies forced cache release and re-raise.

**Comment by hsliuustc0106** [comment]
Self-review on `1843f05c`:

- Traced encoder/DiT/VAE residency, alias-preserving storage rebinding, allocator bounds, OOM retry, failure cleanup, and unchanged executor shutdown ownership.
- Rechecked both component-failure and success-path offload-failure branches; the original exception is preserved and forced cache release is attempted in each path.
- Verified the feature/module docs match the 25% cached and 5% physical-free implementation bounds and document the intentional HBM tradeoff.
- Local validation: 11 focused tests; 339 broader CPU tests with 14 deselected; all changed-file pre-commit hooks passed.
- Matched rebased L20X A/B: `main` 18.317 ± 0.267 s versus head 15.580 ± 0.032 s (**-14.94%**), with byte-identical outputs and clean shutdown. Peak HBM is 20,808 → 22,722 MB as documented.
- InferMatrixCopilot rechecked the final head; all prior findings are fixed and no new actionable finding remained.
